ethcore/src/trace/db.rs

```diff
@@ -0,0 +1,589 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Fat database.
+use std::ptr;
+use std::ops::Deref;
+use std::collections::HashMap;
+use std::sync::{RwLock, Arc};
+use std::path::Path;
+use bloomchain::Number;
+use bloomchain::group::{BloomGroupDatabase, BloomGroupChain, GroupPosition, BloomGroup};
+use util::{FixedHash, H256, H264, Database, DBTransaction};
+use header::BlockNumber;
+use trace::{BlockTraces, LocalizedTrace, Config, Switch, Filter, Database as TraceDatabase, ImportRequest,
+DatabaseExtras};
+use db::{Key, Writable, Readable, CacheUpdatePolicy};
+use super::bloom::{TraceGroupPosition, BlockTracesBloom, BlockTracesBloomGroup};
+use super::flat::{FlatTrace, FlatBlockTraces, FlatTransactionTraces};
+
+const TRACE_DB_VER: &'static [u8] = b"1.0";
+
+#[derive(Debug, Copy, Clone)]
+pub enum TracedbIndex {
+	/// Block traces index.
+	BlockTraces = 0,
+	/// Trace bloom group index.
+	BlockTracesBloomGroups = 1,
+}
+
+impl Key<BlockTraces> for H256 {
+	type Target = H264;
+
+	fn key(&self) -> H264 {
+		let mut result = H264::default();
+		result[0] = TracedbIndex::BlockTraces as u8;
+		unsafe {
+			ptr::copy(self.as_ptr(), result.as_mut_ptr().offset(1), 32);
+		}
+		result
+	}
+}
+
+/// Helper data structure created cause [u8; 6] does not implement Deref to &[u8].
+pub struct TraceGroupKey([u8; 6]);
+
+impl Deref for TraceGroupKey {
+	type Target = [u8];
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl Key<BlockTracesBloomGroup> for TraceGroupPosition {
+	type Target = TraceGroupKey;
+
+	fn key(&self) -> Self::Target {
+		let mut result = [0u8; 6];
+		result[0] = TracedbIndex::BlockTracesBloomGroups as u8;
+		result[1] = self.level;
+		unsafe {
+			ptr::copy(&[self.index] as *const u32 as *const u8, result.as_mut_ptr().offset(2), 4);
+		}
+		TraceGroupKey(result)
+	}
+}
+
+/// Fat database.
```

Comment:
`Fat DB` is generally reserved for storing the prehash images of the storage/state keys. Present functionality is "transaction tracing".